consensus/src/block_storage/block_store.rs

```diff
@@ -219,21 +242,14 @@ impl BlockStore {
             .commit(&blocks_to_commit, finality_proof)
             .await
             .expect("Failed to persist commit");
-        update_counters_for_committed_blocks(&blocks_to_commit);
-        let current_round = self.root().round();
-        let committed_round = block_to_commit.round();
-        debug!(
-            LogSchema::new(LogEvent::CommitViaBlock).round(current_round),
-            committed_round = committed_round,
-            block_id = block_to_commit.id(),
-        );
-        event!("committed",
-            "block_id": block_to_commit.id().short_str(),
-            "epoch": block_to_commit.epoch(),
-            "round": committed_round,
-            "parent_id": block_to_commit.parent_id().short_str(),
-        );
-        self.prune_tree(block_to_commit.id());
+
+        // if decoupled_execution is turned on, we delay the block pruning
+        if self.decoupled_execution {
+            return Ok(())
+        }
+
+        self.update_counters_and_prune_blocks(&blocks_to_commit, block_to_commit).await;
```

Comment:
maybe we can wrap a commit_callback closure and pass to StateComputer.commit. In OrderingStateComputer we can just pass to next phase, in real state computer, it'll invoke after commit succeeds

Reply:
But that will create a closure for every vector of blocks..? How does rust implement temporary closure passing? Does it pass a function address?

Reply:
Maybe we can simply sets a function pointer in state computer when initiating the block storage

Reply:
wrapping the closure with a Box would move it to heap and only pass the pointer

Reply:
resolved